### PR TITLE
Implemented two factor

### PR DESCRIPTION
**Pull request by Superbil**
I just implemented two factor flow for Apple, this will solved https://github.com/fastlane/fastlane/issues/4391

### CODE REVIEW

**Comment by KrauseFx** [comment]
Do you know how I can setup my Apple account to make use of the new 2 step?

**Comment by Superbil** [comment]
Simple flow

1. go to https://appleid.apple.com/account/home close two step.
2. Then use iPhone or iMac to enable it.
Detail flow can read this document https://support.apple.com/en-us/HT204915

File: spaceship/lib/spaceship/two_step_client.rb [hunk 1]
```diff
@@ -36,12 +36,53 @@ def handle_two_step(response)
...
+        req.headers["Accept"] = "application/json"
+        req.headers['Content-Type'] = 'application/json'
+        req.headers["scnt"] = @scnt
+        req.headers["X-Apple-Id-Session-Id"] = @x_apple_id_session_id
+        req.body = { "securityCode" => { "code" => code.to_s } }.to_json
+      end
+
+      # we use `Spaceship::TunesClient.new.handle_itc_response`
+      # since this might be from the Dev Portal, but for 2 step
+      Spaceship::TunesClient.new.handle_itc_response(r.body)
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/two_step_client.rb]: Why do we parse the response twice?

File: spaceship/lib/spaceship/two_step_client.rb [hunk 2]
```diff
@@ -36,12 +36,53 @@ def handle_two_step(response)
...
+      r = request(:post) do |req|
+        req.url "https://idmsa.apple.com/appleauth/auth/verify/trusteddevice/securitycode"
+        req.headers["Accept"] = "application/json"
+        req.headers['Content-Type'] = 'application/json'
+        req.headers["scnt"] = @scnt
+        req.headers["X-Apple-Id-Session-Id"] = @x_apple_id_session_id
+        req.body = { "securityCode" => { "code" => code.to_s } }.to_json
+      end
+
+      # we use `Spaceship::TunesClient.new.handle_itc_response`
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/two_step_client.rb]: Are those comments relevant, or just copied from the other method?

File: spaceship/lib/spaceship/two_step_client.rb [hunk 3]
```diff
@@ -36,12 +36,53 @@ def handle_two_step(response)
...
     end
 
+    def handle_two_fator(response)
+      security_code = response.body["phoneNumberVerification"]["securityCode"]
+      # {"length"=>6,
+      #  "tooManyCodesSent"=>false,
+      #  "tooManyCodesValidated"=>false,
+      #  "securityCodeLocked"=>false}
+      code_length = security_code["length"]
+      puts "Successfully requested notification"
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/two_step_client.rb]: We didn't actually request a notification here, right?

File: spaceship/lib/spaceship/two_step_client.rb [hunk 4]
```diff
@@ -128,6 +169,12 @@ def select_device(r, device_id)
         raise ex
       end
 
+      store_session
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/two_step_client.rb]: Why was this added here?

File: spaceship/lib/spaceship/two_step_client.rb [hunk 5]
```diff
@@ -150,8 +197,6 @@ def select_device(r, device_id)
       # correct DES... cookie
 
       self.store_cookie
-
-      return true
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/two_step_client.rb]: Why was this changed?

File: spaceship/lib/spaceship/two_step_client.rb [hunk 6]
```diff
@@ -36,12 +36,53 @@ def handle_two_step(response)
...
+      end
+
+      # we use `Spaceship::TunesClient.new.handle_itc_response`
+      # since this might be from the Dev Portal, but for 2 step
+      Spaceship::TunesClient.new.handle_itc_response(r.body)
+
+      begin
+        Spaceship::TunesClient.new.handle_itc_response(r.body) # this will fail if the code is invalid
+      rescue => ex
+        if ex.to_s.include?("verification code") # to have a nicer output
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/two_step_client.rb]: Is that actually an error that might occur?

**Comment by KrauseFx** [comment]
I couldn't enable it. I already have the old 2 step enabled, and didn't find out how I can upgrade to the new one.
I added some comments for now 👍

File: spaceship/lib/spaceship/two_step_client.rb (hunk 2 above)
Comment (Superbil) [inline spaceship/lib/spaceship/two_step_client.rb]: Those comments are relevant, it's little different with two-step.
Two-factor use `securityCode`, two-step use `code`

File: spaceship/lib/spaceship/two_step_client.rb (hunk 1 above)
Comment (Superbil) [inline spaceship/lib/spaceship/two_step_client.rb]: I had been to request just one time, but it failed, but two times will success.

File: spaceship/lib/spaceship/two_step_client.rb (hunk 4 above)
Comment (Superbil) [inline spaceship/lib/spaceship/two_step_client.rb]: After everything finish need to store, so little refactoring that to a new function `store_session`. That let `select_device` and `handle_two_fator` to use it.

File: spaceship/lib/spaceship/two_step_client.rb (hunk 5 above)
Comment (Superbil) [inline spaceship/lib/spaceship/two_step_client.rb]: This is the same return value before, `return true` is moved to line:174.

File: spaceship/lib/spaceship/two_step_client.rb (hunk 3 above)
Comment (Superbil) [inline spaceship/lib/spaceship/two_step_client.rb]: Yea, we didn't request a notification, it's an useless information.

**Comment by Superbil** [comment]
If you had 2 step enabled, you can't enable 2 factor in the same time. You must disable 2 step then you can upgrade to the new one.

File: spaceship/lib/spaceship/two_step_client.rb (hunk 6 above)
Comment (Superbil) [inline spaceship/lib/spaceship/two_step_client.rb]: I try it again. That won't happen, I will remove it.

File: spaceship/lib/spaceship/two_step_client.rb (hunk 1 above)
Comment (Superbil) [inline spaceship/lib/spaceship/two_step_client.rb]: I seen parse one was ok, I will remove one.

**Comment by ekurutepe** [comment]
@Superbil @KrauseFx I was able to verify this as working fine with the push based two-factor auth.

**Comment by KrauseFx** [comment]
I disabled and re-enabled 2 step and still have the old 2 factor auth process. I'll just review the code one more time and trust you two on the testing part 👍

**Comment by KrauseFx** [comment]
Thanks again for your great contribution to fastlane @Superbil 🚀 I'll bundle a new release now
